core/authentication_api.php

```diff
@@ -717,6 +717,20 @@ function auth_does_password_match( $p_user_id, $p_test_password ) {
 	}
 
 	$t_password = user_get_field( $p_user_id, 'password' );
+
+	# Process SAFE_HASH separately from the legacy methods, to safeguard against
+	# timing attacks leveraging password_verify() function.
+	if(    SAFE_HASH == $t_configured_login_method
+		&& password_verify( $p_test_password, $t_password )
+	) {
+		# Update the password hash if it does not match the current algorithm
+		if( password_needs_rehash( $t_password, PASSWORD_DEFAULT ) ) {
+			user_set_password( $p_user_id, $p_test_password, true );
```

Comment:
Changing a password *should* cause all sessions to be terminated, in general.

However, in this specific case I agree that this is really a "technical" change (i.e. we have a new hash, but the password itself remains the same) so there is no need to actually log the user out.

Are you suggesting to call `user_set_field( $p_user_id, 'password', $p_test_password )` instead ? 
Note that this would require special handling for protected accounts (since *user_set_fields* will trigger an error in this case). Alternatively, adding a new parameter to user_set_field, or defining a new API function that doesn't reset the cookie / token could work as well.